### PR TITLE
DropShadowFilter initial commit

### PR DESCRIPTION
**Pull request by funkypaul**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,6 +32,7 @@
     "build:shockwave": "pixify -s src/shockwave -n shockwave",
     "build:tiltshift": "pixify -s src/tiltshift -n tiltshift",
     "build:twist": "pixify -s src/twist -n twist",
+    "build:dropshadow": "pixify -s src/dropshadow -n dropshadow",
```
Comment (bigtimebuddy) [inline package.json]: Also needs to be added to the `build` script.

File: pixi-filters.iml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<module type="WEB_MODULE" version="4">
+  <component name="NewModuleRootManager" inherit-compiler-output="true">
+    <exclude-output />
+    <content url="file://$MODULE_DIR$" />
+    <orderEntry type="sourceFolder" forTests="false" />
+  </component>
+</module>
```
Comment (bigtimebuddy) [inline pixi-filters.iml]: Please remove this .iml file. Repo should be IDE agnostic.

File: examples/index.html [hunk 3]
```diff
@@ -44,6 +45,10 @@
             loader.add('pug', 'pug.jpg')
                 .load(function(loader, resources){
                     var texture = resources.pug.texture;
+                    sprite.scale.x = 0.9;
+                    sprite.scale.y = 0.9;
+                    sprite.x = 5;
+                    sprite.y = 5;
```
Comment (bigtimebuddy) [inline examples/index.html]: Rather than scaling down the sprite, we should just make the canvas a little larger than the image.

**Comment by VsevolodZakharov** [comment]
@bigtimebuddy, will DropShadowFilter be included in the next PixiJS release? It is missing in the 4.0.1 too as far as I can see.

**Comment by GoodBoyDigital** [comment]
Life is pretty busy at the mo, but this task is next on my hit list peeps 👍

**Comment by kennardconsulting** [comment]
+1

**Comment by Ansien12** [comment]
@GoodBoyDigital Any updates?

**Comment by alstrup** [comment]
We are also very interested in this. We have been waiting since 4.0. It looks from the review comments that the review requests only deal with examples, while the filter itself works. So maybe this can be considered as is for the next release?

**Comment by ansien12** [comment]
@alstrup I haven't been able to get funkypaul's implementation working. Gives me some undefined errors.

**Comment by funkypaul** [comment]
Its not finished, there are errors in pixi itself preventing from having dropshadow working. I reported that to pixi guys, but it seems theyre pretty busy atm.

**Comment by alstrup** [comment]
Any details if we could have somehow help out?

**Comment by funkypaul** [comment]
If you lanch the branch you can see that:

FilterManager.js:289 
Uncaught TypeError: Cannot read property 'bind' of undefined

the gl.TEXTURE0 is undefined

This is in filtermanager. I didnt proceed from that step.

**Comment by VsevolodZakharov** [comment]
@funkypaul, Hi, there is a problem when building this dropshadow filter with npm. Am I doing something wrong?

npm run build:dropshadow

> pixi-filters@1.0.5 build:dropshadow /Users/vzakharov/pixi-filters-DropShadowFilter
> pixify -s src/dropshadow -n dropshadow

/Users/vzakharov/pixi-filters-DropShadowFilter/node_modules/pixify/index.js:49
    options = Object.assign({}, options, args);
                     ^
TypeError: undefined is not a function
    at bundle (/Users/vzakharov/pixi-filters-DropShadowFilter/node_modules/pixify/index.js:49:22)
    at Object.<anonymous> (/Users/vzakharov/pixi-filters-DropShadowFilter/node_modules/pixify/index.js:61:1)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10)
    at startup (node.js:129:16)
    at node.js:814:3

npm ERR! Darwin 15.6.0
npm ERR! argv "node" "/usr/local/bin/npm" "run" "build:dropshadow"
npm ERR! node v0.12.0
npm ERR! npm  v2.5.1
npm ERR! code ELIFECYCLE
npm ERR! pixi-filters@1.0.5 build:dropshadow: `pixify -s src/dropshadow -n dropshadow`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the pixi-filters@1.0.5 build:dropshadow script 'pixify -s src/dropshadow -n dropshadow'.
npm ERR! This is most likely a problem with the pixi-filters package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     pixify -s src/dropshadow -n dropshadow
npm ERR! You can get their info via:
npm ERR!     npm owner ls pixi-filters
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/vzakharov/pixi-filters-DropShadowFilter/npm-debug.log

**Comment by themoonrat** [comment]
You need to be using node v6.x; you are on 0.12.x

**Comment by VsevolodZakharov** [comment]
@themoonrat , thanks a lot. Builds fine with v6. Unfortunately mac brew does not provide the latest node version.

**Comment by VsevolodZakharov** [comment]
To me that was enough to add a check for default value of sampler2D in the FilterManager.prototype.syncUniforms and do nothing for default value 0. 
...
 //TODO Cacheing layer..
    for(var i in uniformData)
    {
        if(uniformData[i].type === 'sampler2D' && uniforms[i] != 0)
        {
            shader.uniforms[i] = textureCount;
...

**Comment by funkypaul** [comment]
And its appplying shadow correctly? can you share a screenshot? if yes, please create a PR to pixi with this fix.

**Comment by VsevolodZakharov** [comment]
screen shot 2016-10-19 at 12 40 48

That is your DropShadow around rectangle 100x100

**Comment by VsevolodZakharov** [comment]
@funkypaul , minor issue: in the alpha getter/setter this.blurYTintFilter.uniforms.alpha should be used instead of this.blurYTintFilter.uniforms.alpha.value.

**Comment by ronanquillevere** [comment]
Any news on this ?

**Comment by kittykatattack** [comment]
Please, dear Pixi, we need DropShadowFilter! 🙏

**Comment by VsevolodZakharov** [comment]
I used this for Pixi 4:
`DropShadowFilter.sharedCopyFilter = new PIXI.Filter();    

function DropShadowFilter(angle, distance, spread, color, alpha) {
    PIXI.Filter.call(this);

    angle *= Math.PI / 180;
    this.angle = angle;
    this.padding = distance;
    this.distance = distance;
    this.color = color;
    this.padding = distance < 10 ? 10 : distance;
    this.blur = spread;
    this.alpha = alpha;

    this.tintFilter = new PIXI.Filter(
        PIXI.Filter.defaultVertexSrc,
        ['varying vec2 vTextureCoord;',
        'uniform sampler2D uSampler;',
        'uniform float alpha;',
        'uniform vec3 color;',
        'void main(void){',
        '   vec4 sample = texture2D(uSampler, vTextureCoord);',
        '   gl_FragColor = vec4(color, sample.a > 0.0 ? alpha : 0.0);',
      '}'].join("\n")
    );
    this.tintFilter.uniforms.alpha = alpha;
    this.tintFilter.uniforms.color = PIXI.utils.hex2rgb(color);

    this.blurFilter = new PIXI.filters.BlurFilter();
    this.blurFilter.blur = spread;
}

DropShadowFilter.prototype = Object.create(PIXI.Filter.prototype);
DropShadowFilter.prototype.constructor = DropShadowFilter;
DropShadowFilter.prototype.apply = function (filterManager, input, output) {
    var rt = filterManager.getRenderTarget();
    rt.clear();
    if (!output.root) output.clear();

    rt.transform = new PIXI.Matrix();
    rt.transform.translate(this.distance * Math.cos(this.angle), this.distance * Math.sin(this.angle));
    this.tintFilter.apply(filterManager, input, rt);
    this.blurFilter.apply(filterManager, rt, output);
    DropShadowFilter.sharedCopyFilter.apply(filterManager, input, output);

    rt.transform = null;
    filterManager.returnRenderTarget(rt);
};

PIXI.filters.DropShadowFilter = DropShadowFilter;`

**Comment by kennardconsulting** [comment]
@VsevolodZakharov thanks for this! However it doesn't quite work? It appears to be incorrectly copy/pasted from your other post here: http://www.html5gamedevs.com/topic/24999-dropshadowfilter/

And even when using that other post, the shadow always appears white? How would I get a black shadow?

**Comment by VsevolodZakharov** [comment]
@kennardconsulting , it is not copy-paste , the version above is better (no bugs, working shadow parameters). Unfortunately GitHub code tag does not work well for this piece of code somehow. I've attached the file here - just include that after including pixi.min.js 
[pixi.filters.js.zip](https://github.com/pixijs/pixi-filters/files/738328/pixi.filters.js.zip). 
Filter constructor example is "new DropShadowFilter(0.5, 10.0, 2.0,  0, 1.0)"

**Comment by ansien12** [comment]
Just upgraded to 4.0 after two years, can confirm that @VsevolodZakharov's solution works perfectly!

**Comment by bigtimebuddy** [comment]
Cleaned up a version of DropShowFilter posted here by @VsevolodZakharov. Will add to coming v1.1, see PR #30.

**Comment by kennardconsulting** [comment]
Fantastic! Thanks everyone!

**Pull request closed without merging**